### PR TITLE
Setup command to generate report widgets

### PR DESCRIPTION
**Pull request by LarryBarker**
This PR implements a command to generate report widgets, which can help streamline the development process. The command to generate a report widget is `php artisan create:reportwidget Author.Plugin ReportWidgetName`. 

The functionality is very similar to the scaffolding for form widgets, with the exception that save values are not needed. 

The view markup was based on one of the report widgets from the RainLab Google Analytics report widgets. The generated widget scaffolds a function to prepare view variables, and throws an exception if any errors occur. There is also a default report widget property for the title. 

My tests locally are working.

### CODE REVIEW

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /**
+     * @inheritDoc
+     */
+    public function render()
+    {
+        try {
+            $this->loadData();
+        }
+        catch (Exception $ex) {
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:18]:
```suggestion
        } catch (Exception $ex) {
```

File: src/Scaffold/ScaffoldServiceProvider.php [hunk 2]
```diff
@@ -40,12 +40,17 @@ public function register()
             return new CreateCommand;
         });
 
+        $this->app->singleton('command.create.reportwidget', function () {
+            return new CreateReportWidget;
+        });
```
Comment (LukeTowers) [inline src/Scaffold/ScaffoldServiceProvider.php:45]: Should go after CreateFormWidget

File: src/Scaffold/ScaffoldServiceProvider.php [hunk 3]
```diff
@@ -40,12 +40,17 @@ public function register()
...
+
         $this->commands('command.create.plugin');
         $this->commands('command.create.model');
         $this->commands('command.create.controller');
         $this->commands('command.create.component');
         $this->commands('command.create.formwidget');
         $this->commands('command.create.command');
+        $this->commands('command.create.reportwidget');
```
Comment (LukeTowers) [inline src/Scaffold/ScaffoldServiceProvider.php:53]: Should go after create form widget

File: src/Scaffold/ScaffoldServiceProvider.php [hunk 4]
```diff
@@ -60,7 +65,8 @@ public function provides()
             'command.create.controller',
             'command.create.component',
             'command.create.formwidget',
-            'command.create.command'
+            'command.create.command',
+            'command.create.reportwidget',
```
Comment (LukeTowers) [inline src/Scaffold/ScaffoldServiceProvider.php:69]: Should go after create form widget

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return [
+            'title' => [
+                'title'             => 'backend::lang.dashboard.widget_title_label',
+                'default'           => '{{title_name}} Report Widget',
+                'type'              => 'string',
+                'validationPattern' => '^.+$',
+                'validationMessage' => 'backend::lang.dashboard.widget_title_error'
+            ];
+    }
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:45]:
```suggestion
            ],
        ];
    }
```

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public function defineProperties()
+    {
+        return [
+            'title' => [
+                'title'             => 'backend::lang.dashboard.widget_title_label',
+                'default'           => '{{title_name}} Report Widget',
+                'type'              => 'string',
+                'validationPattern' => '^.+$',
+                'validationMessage' => 'backend::lang.dashboard.widget_title_error'
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:42]:
```suggestion
                'validationMessage' => 'backend::lang.dashboard.widget_title_error',
```

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ */
+class {{studly_name}} extends ReportWidgetBase
+{
+    /**
+     * @inheritDoc
+     */
+    public function render()
+    {
+        try {
+            $this->loadData();
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:16]:
```suggestion
            $this->prepareVars();
```

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * {{name}} Report Widget
+ */
+class {{studly_name}} extends ReportWidgetBase
+{
+    /**
+     * @inheritDoc
+     */
+    public function render()
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:13]:
```suggestion
    /**
     * @var string The default alias to use for this widget
     */
    protected $defaultAlias = '{{studly_name}}ReportWidget';

    /**
     * Defines the widget's properties
     * @return array
     */
    public function defineProperties()
    {
        return [
            'title' => [
                'title'             => 'backend::lang.dashboard.widget_title_label',
                'default'           => '{{title_name}} Report Widget',
                'type'              => 'string',
                'validationPattern' => '^.+$',
                'validationMessage' => 'backend::lang.dashboard.widget_title_error',
            ],
        ];
    }
    
    /**
     * Adds widget specific asset files. Use $this->addJs() and $this->addCss()
     * to register new assets to include on the page.
     * @return void
     */
    protected function loadAssets()
    {
    }
    
    /**
     * Renders the widget's primary contents.
     * @return string HTML markup supplied by this widget.
     */
    public function render()
```

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        return [
+            'title' => [
+                'title'             => 'backend::lang.dashboard.widget_title_label',
+                'default'           => '{{title_name}} Report Widget',
+                'type'              => 'string',
+                'validationPattern' => '^.+$',
+                'validationMessage' => 'backend::lang.dashboard.widget_title_error',
+            ],
+        ];
+    }
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:45]:
```suggestion
```

File: src/Scaffold/Console/reportwidget/formwidget.stub [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        return $this->makePartial('{{lower_name}}');
+    }
+
+    /**
+     * Prepares the report widget view data
+     */
+    public function prepareVars()
+    {
+    }
+
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/formwidget.stub:62]:
```suggestion
```

**Comment by LukeTowers** [comment]
Please also verify that the generated report widget can be added to the dashboard and works properly and can have its properties changed.

**Comment by LarryBarker** [comment]
Ok, the errors in the stub file are fixed and I have tested creating a widget and adding it to the dashboard.
image
image

File: src/Scaffold/Console/CreateReportWidget.php [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    protected $type = 'ReportWidget';
+
+    /**
+     * A mapping of stub to generated file.
+     *
+     * @var array
+     */
+    protected $stubs = [
+        'reportwidget/reportwidget.stub'      => 'reportwidgets/{{studly_name}}.php',
+        'reportwidget/widget.stub'         => 'reportwidgets/{{lower_name}}/partials/_{{lower_name}}.htm',
```
Comment (LukeTowers) [inline src/Scaffold/Console/CreateReportWidget.php:37]:
```suggestion
        'reportwidget/reportwidget.stub' => 'reportwidgets/{{studly_name}}.php',
        'reportwidget/widget.stub'       => 'reportwidgets/{{lower_name}}/partials/_{{lower_name}}.htm',
```

File: src/Scaffold/Console/reportwidget/widget.stub [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<div class="report-widget">
+    <h3><?= e($this->property('title')) ?></h3>
+
+    <?php if (!isset($error)): ?>
+
```
Comment (LukeTowers) [inline src/Scaffold/Console/reportwidget/widget.stub:5]:
```suggestion
        <p>This is the default partial content.</p>
```

**Pull request merged**